Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -140,43 +138,73 @@ namespace ROS2
         message.header.stamp = ROS2Interface::Get()->GetROSTimestamp();
         message.height = 1;
-        message.width = lastScanResults.m_points.size();
-        message.point_step = sizeof(AZ::Vector3);
+        message.width = results.m_points.size();
+        message.point_step = sizeof(AZ::Vector3) + (isIntensityEnabled ? sizeof(float) : 0U);
         message.row_step = message.width * message.point_step;
 
-        AZStd::array<const char*, 3> point_field_names = { "x", "y", "z" };
-        for (int i = 0; i < point_field_names.size(); i++)
+        AZStd::array point_field_names = { "x", "y", "z", "intensity" };
+        const size_t numPointFields = isIntensityEnabled ? point_field_names.size() : point_field_names.size() - 1U;
+        for (int i = 0; i < numPointFields; i++)
         {
             sensor_msgs::msg::PointField pf;
             pf.name = point_field_names[i];
-            pf.offset = i * 4;
+            pf.offset = i * sizeof(float);
+            if (i == point_field_names.size() - 1U)
+            { // AZ::Vector3 may have padding.
+                pf.offset = sizeof(AZ::Vector3);
+            }
+
             pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
             pf.count = 1;
             message.fields.push_back(pf);
         }
 
-        size_t sizeInBytes = lastScanResults.m_points.size() * sizeof(AZ::Vector3);
+        m_rosMessageData.clear();
+        const size_t sizeInBytes = results.m_points.size() * message.point_step;
         message.data.resize(sizeInBytes);
+        m_rosMessageData.reserve(sizeInBytes);
+
+        AZ_Assert(
+            !isIntensityEnabled || (results.m_points.size() == results.m_intensity.size()),
+            "Inconsistency in the size of raycast result attributes.");
+
+        const auto entityTransform = GetEntity()->FindComponent<AzFramework::TransformComponent>();
+        const auto inverseLidarTM = entityTransform->GetWorldTM().GetInverse();
+        for (size_t pointIdx = 0; pointIdx < results.m_points.size(); ++pointIdx)
+        {
+            const AZ::Vector3 globalPoint = inverseLidarTM.TransformPoint(results.m_points[pointIdx]);
+            const auto positionBytes = reinterpret_cast<const uint8_t*>(&globalPoint);
+            m_rosMessageData.insert(m_rosMessageData.end(), positionBytes, positionBytes + sizeof(AZ::Vector3));
+
+            if (isIntensityEnabled)
+            {
+                float intensity = results.m_intensity[pointIdx];
+                const auto intensityBytes = reinterpret_cast<const uint8_t*>(&intensity);
+                m_rosMessageData.insert(m_rosMessageData.end(), intensityBytes, intensityBytes + sizeof(float));
+            }
+        }
+
         AZ_Assert(message.row_step * message.height == sizeInBytes, "Inconsistency in the size of point cloud data");
-        memcpy(message.data.data(), lastScanResults.m_points.data(), sizeInBytes);
+        memcpy(message.data.data(), m_rosMessageData.data(), sizeInBytes);
```

Comment:
This looks like unnecessary step - you copy data to `m_rosMessageData` and then to [message.data](url) instead of omitting this step. What is the reason behind this? I mean, instead of insertions above you can just memcpy part by part into `message.data`.

Reply:
iterators solve this beautifully.